File: backend/app/schemas/club.py
from datetime import date, datetime

from pydantic import BaseModel

from app.models.profile import UserRole
from app.models.season import SeasonStatus


# ── Club ──────────────────────────────────────────────────────────────────────

class ClubCreate(BaseModel):
    name: str


class ClubUpdate(BaseModel):
    name: str | None = None
    logo_url: str | None = None


class ClubResponse(BaseModel):
    id: int
    name: str
    logo_url: str | None
    archived_at: datetime | None
    created_at: datetime

    model_config = {"from_attributes": True}


# ── Season ────────────────────────────────────────────────────────────────────

class SeasonCreate(BaseModel):
    name: str
    starts_at: date | None = None
    ends_at: date | None = None


class SeasonStatusUpdate(BaseModel):
    status: SeasonStatus


class SeasonResponse(BaseModel):
    id: int
    club_id: int
    name: str
    status: SeasonStatus
    starts_at: date | None
    ends_at: date | None
    created_at: datetime

    model_config = {"from_attributes": True}


# ── Team ──────────────────────────────────────────────────────────────────────

class TeamCreate(BaseModel):
    name: str
    season_id: int


class TeamResponse(BaseModel):
    id: int
    club_id: int
    season_id: int
    name: str
    archived_at: datetime | None
    created_at: datetime

    model_config = {"from_attributes": True}


# ── ClubMember ────────────────────────────────────────────────────────────────

class AddMemberRequest(BaseModel):
    """Añadir un usuario existente como miembro del club (RF-041)."""
    user_id: int


class ClubMemberResponse(BaseModel):
    id: int
    club_id: int
    user_id: int
    invited_by: int | None
    joined_at: datetime
    archived_at: datetime | None

    model_config = {"from_attributes": True}


# ── Profile ───────────────────────────────────────────────────────────────────

class AssignProfileRequest(BaseModel):
    """Asignar un perfil (rol en equipo/club) a un miembro del club."""
    user_id: int
    role: UserRole
    season_id: int
    team_id: int | None = None  # NULL para TechnicalDirector


class ProfileResponse(BaseModel):
    id: int
    user_id: int
    club_id: int
    team_id: int | None
    season_id: int
    role: UserRole
    archived_at: datetime | None
    created_at: datetime
    # Campos enriquecidos para el selector de perfil (cargados en el router)
    club_name: str | None = None
    team_name: str | None = None
    season_name: str | None = None

    model_config = {"from_attributes": True}

File: shared/types/club.ts
// ── Enums ─────────────────────────────────────────────────────────────────────

export type SeasonStatus = "future" | "active" | "archived";

export type UserRole = "technical_director" | "head_coach" | "staff_member";

// ── Club ──────────────────────────────────────────────────────────────────────

export interface Club {
  id: number;
  name: string;
  logo_url: string | null;
  archived_at: string | null;
  created_at: string;
}

export interface ClubCreate {
  name: string;
}

// ── Season ────────────────────────────────────────────────────────────────────

export interface Season {
  id: number;
  club_id: number;
  name: string;
  status: SeasonStatus;
  starts_at: string | null;
  ends_at: string | null;
  created_at: string;
}

export interface SeasonCreate {
  name: string;
  starts_at?: string;
  ends_at?: string;
}

// ── Team ──────────────────────────────────────────────────────────────────────

export interface Team {
  id: number;
  club_id: number;
  season_id: number;
  name: string;
  archived_at: string | null;
  created_at: string;
}

export interface TeamCreate {
  name: string;
  season_id: number;
}

// ── ClubMember ────────────────────────────────────────────────────────────────

export interface ClubMember {
  id: number;
  club_id: number;
  user_id: number;
  user_email: string | null;
  invited_by: number | null;
  joined_at: string;
  archived_at: string | null;
}

// ── Profile ───────────────────────────────────────────────────────────────────

export interface Profile {
  id: number;
  user_id: number;
  club_id: number;
  team_id: number | null;
  season_id: number;
  role: UserRole;
  archived_at: string | null;
  created_at: string;
  /** Campos enriquecidos para el selector de perfil */
  club_name: string | null;
  team_name: string | null;
  season_name: string | null;
  user_email: string | null;
}

/** Etiqueta legible para mostrar en el selector de perfil */
export function profileLabel(profile: Profile): string {
  const roleLabel: Record<UserRole, string> = {
    technical_director: "Director técnico",
    head_coach: "Entrenador",
    staff_member: "Cuerpo técnico",
  };
  const parts = [roleLabel[profile.role]];
  if (profile.team_name) parts.push(profile.team_name);
  if (profile.club_name) parts.push(profile.club_name);
  if (profile.season_name) parts.push(profile.season_name);
  return parts.join(" — ");
}
